File: 1. Subarray Sum Equals K.py
"""
Time Complexity = O(n) because we traverse the array only once.
Space Complexity = O(n) for the map"""


class Solution:
    def subarraySum(self, nums, k: int) -> int:
        result = 0
        running_sum = 0
        table = {}
        # Initializing the table
        table[0] = 1

        # Traversing the array
        for num in nums:
            # Calculating running sum
            running_sum += num

            # If r_sum - k is already in the table, then add its value to the result
            if (running_sum - k) in table:
                result += table[running_sum - k]

            # If already present, increment the count
            if running_sum in table:
                table[running_sum] += 1
            # If not present, initialize it to 1
            else:
                table[running_sum] = 1

        return result

File: 1.subarray-sum-equals-k.java
// Time Complexity :O(n)
// Space Complexity :o(n)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No
class Solution {
    public int subarraySum(int[] nums, int k) {

        HashMap<Integer , Integer> map = new HashMap<>();
        int rsum=0, count=0;
        map.put(0,1);
        //recurring sum and Count value

        for(int i=0;i<nums.length; i++)
        {
            rsum= rsum+ nums[i];
            //if (recurring sum - k) already present
            if(map.containsKey(rsum-k))
            {
                count=count+map.get(rsum-k);
            }

            if(map.containsKey(rsum))
            {
                map.put(rsum , map.get(rsum)+1);
            }
            else{
                map.put(rsum, map.getOrDefault(rsum,0)+1);
            }
        }
        return count;
    }
}

// Your code here along with comments explaining your approach
